File: utilslib/src/main/java/com/ashlikun/utils/ui/ActivityManager.java
package com.ashlikun.utils.ui;

import android.app.Activity;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Stack;

/**
 * 作者　　: 李坤
 * 创建时间: 2017/3/27 17:54
 * <p>
 * 方法功能：存放Activity的栈
 */

public class ActivityManager {
    private static Stack<Activity> activityStack = new Stack<>();
    private static ActivityManager instance;

    /**
     * 获取指定的运行中的activity,只取最新的
     */
    public <T> T getTagActivity(Class<? extends Activity> activity) {
        Activity returnAct = null;
        if (activity != null) {
            for (Activity a : activityStack) {
                if (a.getClass() == activity) {
                    returnAct = a;
                    break;
                }
            }
        }
        if (returnAct != null && returnAct.isFinishing()) {
            activityStack.remove(returnAct);
            returnAct = null;
            return getTagActivity(activity);
        }
        return (T) returnAct;
    }

    /**
     * 获取指定的运行中的activity，取全部的
     */
    public <T extends Activity> List<T> getTagActivitys(Class<? extends Activity> activity) {
        List<T> returnAct = new ArrayList<T>();
        if (activity != null) {
            for (Activity a : activityStack) {
                if (a.getClass() == activity) {
                    returnAct.add((T) a);
                }
            }
        }
        Iterator iterator = returnAct.iterator();
        while (iterator.hasNext()){

        }
        for (int i = 0; i < ; i++) {

        }
        for (T a : returnAct) {
            if (a != null && a.isFinishing()) {
                activityStack.remove(a);
            }
        }

        return returnAct;
    }

    private ActivityManager() {
    }

    /**
     * 获取前台 Activity
     */
    public static Activity getForegroundActivity() {
        return getInstance().currentActivity();
    }

    public static ActivityManager getInstance() {
        if (instance == null) {
            instance = new ActivityManager();
        }
        return instance;
    }

    /**
     * 退出Activity
     *
     * @param activity
     */
    public void exitActivity(Activity activity) {
        if (activity != null) {
            if (activityStack.contains(activity)) {
                // 在从自定义集合中取出当前Activity时，也进行了Activity的关闭操作
                activityStack.remove(activity);
                activity.finish();
                activity = null;
            }
        }
    }

    /**
     * 退出Activity
     *
     * @param activity
     */
    public void exitActivity(Class activity) {
        if (activity != null) {
            // 在从自定义集合中取出当前Activity时，也进行了Activity的关闭操作
            for (Activity a : activityStack) {
                if (a.getClass() == activity) {
                    activityStack.remove(a);
                    a.finish();
                    a = null;
                    break;
                }
            }

        }
    }

    /**
     * 是否已经有这个activity
     *
     * @param activity
     * @return
     */
    public Activity contains(Class<? extends Activity> activity) {
        if (!activityStack.isEmpty()) {
            for (int i = 0; i < activityStack.size(); i++) {
                if (!activityStack.get(i).isFinishing() &&
                        activityStack.get(i).getClass() == activity) {
                    return activityStack.get(i);
                }
            }
        }
        return null;
    }

    /**
     * 回退到指定的activity
     */
    public boolean goBack(Class<? extends Activity> activity) {
        if (contains(activity) != null) {
            while (activityStack.size() > 1) {
                if (!currentActivity(activity)) {
                    exitTopActivity();
                } else {
                    break;
                }
            }
            return true;

        }
        return false;
    }

    /**
     * 回退到指定的activity，并且销毁指定的activity
     */
    public boolean goBackFinish(Class<? extends Activity> activity) {
        Activity tagActivity = contains(activity);
        if (tagActivity != null) {
            while (!activityStack.isEmpty()) {
                if (!currentActivity(activity)) {
                    exitTopActivity();
                } else {
                    break;
                }
            }
            //退出当前activity
            exitTopActivity();
            return true;
        }
        return false;
    }

    /**
     * 获得当前栈顶Activity
     *
     * @return
     */
    public Activity currentActivity() {
        Activity activity = null;
        if (activityStack != null && !activityStack.empty()) {
            activity = activityStack.lastElement();
            if (activity != null && activity.isFinishing()) {
                activityStack.remove(activity);
                activity = null;
                return currentActivity();
            }
        }
        return activity;
    }

    /**
     * 当前栈顶是不是这个activity
     *
     * @return
     */
    public boolean currentActivity(Class<? extends Activity> activityClas) {
        Activity activity = currentActivity();
        if (activity != null && activity.getClass() == activityClas) {
            return true;
        }
        return false;
    }

    /**
     * 将当前Activity推入栈中
     *
     * @param activity
     */
    public void pushActivity(Activity activity) {
        if (activityStack == null) {
            activityStack = new Stack<Activity>();
        }
        activityStack.addElement(activity);
    }

    /**
     * 退出栈中所有Activity
     */
    public void exitAllActivity() {

        try {
            while (true) {
                Activity activity = currentActivity();
                if (activity == null) {
                    break;
                }
                exitActivity(activity);
            }
        } catch (Exception e) {

        }

    }

    /**
     * 退出栈中所有Activity
     *
     * @param activity
     */
    public void exitAllActivity(Class... activity) {
        try {
            for (Activity a : activityStack) {
                if (a == null) {
                    continue;
                }
                boolean isfinish = true;
                for (Class c : activity) {
                    if (a.getClass() == c) {
                        isfinish = false;
                        break;
                    }
                }
                if (isfinish) {
                    exitActivity(a);
                }
            }
        } catch (Exception e) {

        }

    }

    /**
     * 退出栈顶activity
     */
    public void exitTopActivity() {
        if (activityStack != null && !activityStack.isEmpty()) {
            Activity activity = activityStack.pop();
            if (activity != null) {
                activity.finish();
            }
        }
    }

    /**
     * 获取 指定activity 个数
     *
     * @return
     */
    public int getActivitySize(Class<? extends Activity> activityClas) {
        int size = 0;
        if (activityStack != null && !activityStack.isEmpty()) {
            ArrayList<Activity> finifshA = new ArrayList<>();
            for (int i = 0; i < activityStack.size(); i++) {
                Activity activity = activityStack.get(i);
                if (activity.isFinishing()) {
                    finifshA.add(activity);
                } else if (activity.getClass() == activityClas) {
                    size++;
                }
            }
            for (Activity a : finifshA) {
                a.finish();
            }
        }
        return size;
    }

    /**
     * 获取activity 个数
     *
     * @return
     */
    public int getActivitySize() {
        if (activityStack != null) {
            return activityStack.size();
        }
        return 0;
    }
}

File: utilslib/src/main/java/com/ashlikun/utils/ui/ActivityManager.kt
package com.ashlikun.utils.ui

import android.app.Activity
import androidx.activity.ComponentActivity
import androidx.fragment.app.FragmentActivity
import com.ashlikun.utils.ui.extend.lastElementOrNull
import com.ashlikun.utils.ui.extend.popOrNull
import java.util.*

/**
 * @author　　: 李坤
 * 创建时间: 2021.12.14 18:18
 * 邮箱　　：496546144@qq.com
 *
 * 功能介绍：存放Activity的栈
 */
/**
 * 获取前台 Activity
 */
val fActivity: Activity?
    get() = ActivityManager.get().currentActivity()

/**
 * 获取前台 Activity
 */
val fCActivity: ComponentActivity?
    get() = ActivityManager.get().getTagActivity(ComponentActivity::class.java)

val fFActivity: FragmentActivity?
    get() = ActivityManager.get().getTagActivity(FragmentActivity::class.java)

inline fun <reified T : Activity> fTagActivity(): T? {
    return ActivityManager.get().getTagActivity(T::class.java)
}

class ActivityManager private constructor() {
    companion object {
        private var activityStack: Stack<Activity> = Stack()
        private val instance by lazy { ActivityManager() }
        fun get(): ActivityManager = instance

        /**
         * 获取前台 Activity
         */
        val foregroundActivity: Activity?
            get() = get().currentActivity()

        /**
         * 获取前台 Activity
         */
        val fActivity: Activity?
            get() = get().currentActivity()

        /**
         * 获取前台 Activity
         */
        val fCActivity: ComponentActivity?
            get() = get().getTagActivity(ComponentActivity::class.java)

        val fFActivity: FragmentActivity?
            get() = get().getTagActivity(FragmentActivity::class.java)

    }

    /**
     * 获取当前全部Activity
     */
    fun getActivitys() = activityStack.toList()

    /**
     * 获取指定的运行中的activity,只取最新的
     */
    fun <T : Activity> getTagActivity(activity: Class<out Activity>?): T? {
        if (activity == null) return null
        var returnAct = activityStack.findLast { activity.isAssignableFrom(it.javaClass) }
        if (returnAct != null && returnAct.isFinishing) {
            activityStack.remove(returnAct)
            return getTagActivity(activity)
        }
        return returnAct as T?
    }

    /**
     * 获取指定的运行中的activity，取全部的
     */
    fun <T : Activity> getTagActivitys(activity: Class<out Activity>?): List<T> {
        if (activity == null) return listOf()
        val returnAct = activityStack.filter {
            activity.isAssignableFrom(it.javaClass)
        }.toMutableList().filter {
            if (it.isFinishing) {
                activityStack.remove(it)
            }
            !it.isFinishing
        }
        return returnAct as List<T>
    }

    /**
     * 退出Activity
     */
    fun exitActivity(activity: Activity?) {
        if (activity == null) return
        // 在从自定义集合中取出当前Activity时，也进行了Activity的关闭操作
        if (activityStack.remove(activity)) {
            if (!activity.isFinishing) {
                activity.finish()
            }
        }
    }

    /**
     * 退出这个Activity
     * @param isOne 如果多个 是否只是退出最新的一个
     * @param activity
     */
    fun exitActivity(activity: Class<out Activity>?, isOne: Boolean = true) {
        if (activity == null) return
        // 在从自定义集合中取出当前Activity时，也进行了Activity的关闭操作
        activityStack.reversed().forEach {
            if (it.javaClass == activity) {
                activityStack.remove(it)
                it.finish()
                if (isOne) return@forEach
            }
        }
    }

    /**
     * 是否已经有这个activity
     *
     * @param activity
     * @return
     */
    fun contains(activity: Class<out Activity>?) =
        if (activity == null) false else activityStack.find {
            !it.isFinishing && it.javaClass == activity
        }

    /**
     * 回退到指定的activity
     * @param isFinishThis 是否销毁这个Activity
     */
    fun goBack(activity: Class<out Activity>?, isFinishThis: Boolean = false): Boolean {
        if (activity == null) return false
        if (contains(activity) != null) {
            while (activityStack.size > 1) {
                if (!currentActivity(activity)) exitTopActivity()
                else break
            }
            //退出当前activity
            if (isFinishThis) exitTopActivity()
            return true
        }
        return false
    }

    /**
     * 获得当前栈顶Activity
     * @param ignoreActivitys 忽略的Activity
     */
    fun currentActivity(ignoreActivitys: Array<Class<Activity>> = emptyArray()): Activity? {
        var activity = if (ignoreActivitys.isEmpty()) activityStack.lastElementOrNull()
        else activityStack.lastOrNull { !ignoreActivitys.contains(it.javaClass) }
        if (activity?.isFinishing == true) {
            activityStack.remove(activity)
            return currentActivity()
        }
        return activity
    }


    /**
     * 当前栈顶是不是这个activity
     *
     * @return
     */
    fun currentActivity(activityClas: Class<out Activity>?) = if (activityClas == null) false else
        currentActivity()?.javaClass == activityClas

    /**
     * 将当前Activity推入栈中
     */
    fun pushActivity(activity: Activity?) {
        if (activity == null) return
        activityStack.addElement(activity)
    }

    /**
     * 将当前Activity从栈中退出
     */
    fun removeActivity(activity: Activity?): Boolean {
        if (activity == null) return false
        // 在从自定义集合中取出当前Activity时，也进行了Activity的关闭操作
        return activityStack.remove(activity)
    }

    /**
     * 退出栈中所有Activity
     */
    fun exitAllActivity() {
        while (true) {
            val activity = currentActivity() ?: break
            exitActivity(activity)
        }
    }

    /**
     * 退出栈中所有Activity
     */
    fun exitAllActivity(vararg activity: Class<out Activity>?) {
        if (activity == null) return
        for (c in activity) {
            exitActivity(c)
        }
    }

    /**
     * 退出栈顶activity
     */
    fun exitTopActivity() {
        activityStack.popOrNull()?.finish()
    }

    /**
     * 获取 指定activity 个数
     */
    fun getActivitySize(activityClas: Class<out Activity>?): Int {
        if (activityClas == null) return 0
        return activityStack.reversed().filter {
            if (it.isFinishing) activityStack.remove(it)
            it.javaClass == activityClas
        }.size
    }

    /**
     * 获取activity 个数
     *
     * @return
     */
    val activitySize: Int
        get() = activityStack.size


}